File: src/master-data-reconciliation/src/main/java/io/ioka/demo/reconcile/CompetingWorkers.java
// SPDX-License-Identifier: Apache-2.0
package io.ioka.demo.reconcile;

import io.ioka.munarium.client.errors.HeadConflictException;
import io.ioka.munarium.client.model.Ledger;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Map;
import java.util.UUID;

/** A dedicated ledger version and a filesystem barrier shared by two containers. */
public final class CompetingWorkers {
    private CompetingWorkers() {}

    private static void await(Path path) throws Exception {
        long deadline = System.nanoTime() + 60_000_000_000L;
        while (!Files.exists(path)) {
            if (System.nanoTime() > deadline) throw new IllegalStateException("Worker barrier timed out");
            Thread.sleep(50);
        }
    }

    public static void run(String action, Path work) throws Exception {
        Files.createDirectories(work);
        try (var client = Bootstrap.ops(false)) {
            if (action.equals("init")) {
                if (Files.exists(work.resolve("version.json"))) throw new IllegalStateException("Fresh race state required");
                FilesUtil.write(work.resolve("version.json"), FilesUtil.json(Map.of("version", client.commands.createVersion())));
                return;
            }
            String version = FilesUtil.read(work.resolve("version.json")).path("version").asText();
            if (action.equals("verify")) {
                var a = FilesUtil.read(work.resolve("a-done.json"));
                var b = FilesUtil.read(work.resolve("b-done.json"));
                int conflicts = a.path("conflicts").asInt() + b.path("conflicts").asInt();
                if (conflicts != 1 || client.query.head(version) != 2 || !a.path("replay_stable").asBoolean() || !b.path("replay_stable").asBoolean()) {
                    throw new IllegalStateException("Expected one conflict, two writes, and stable completed replays");
                }
                FilesUtil.write(work.resolve("quality.json"), FilesUtil.json(Map.of("passed", true, "tests", 1, "conflicts", conflicts, "workers", 2, "version", version)));
                FilesUtil.write(work.resolve("tests.xml"), "<testsuite name=\"competing-worker-containers\" tests=\"1\" failures=\"0\" skipped=\"0\"><testcase name=\"fresh-head-and-key-after-conflict\"/></testsuite>");
                return;
            }
            if (!action.equals("a") && !action.equals("b")) throw new IllegalArgumentException("Unknown worker");
            String peer = action.equals("a") ? "b" : "a";
            long head = client.query.head(version);
            if (head != 0) throw new IllegalStateException("Fresh ledger required");
            FilesUtil.write(work.resolve(action + "-ready.json"), FilesUtil.json(Map.of("head", head)));
            await(work.resolve(peer + "-ready.json"));
            var claim = Ledger.ClaimInput.fact("worker_" + action, "review_status", "reviewed");
            var attempts = new ArrayList<Map<String, Object>>();
            int conflicts = 0;
            for (int attempt = 0; attempt < 2; attempt++) {
                String key = UUID.randomUUID().toString();
                attempts.add(Map.of("expected_head", head, "idempotency_key", key, "claim", claim));
                FilesUtil.write(work.resolve(action + "-journal.json"), FilesUtil.json(attempts));
                try {
                    var result = client.commands.proposeClaim(version, claim, head, key);
                    if (result.isDisputed()) throw new IllegalStateException("Unexpected disputed race claim");
                    FilesUtil.write(work.resolve(action + "-accepted.json"), FilesUtil.json(Map.of("claim_id", result.claim().id())));
                    await(work.resolve(peer + "-accepted.json"));
                    long completedHead = client.query.head(version);
                    var replay = client.commands.proposeClaim(version, claim, head, key);
                    if (!replay.claim().id().equals(result.claim().id()) || client.query.head(version) != completedHead) throw new IllegalStateException("Completed replay changed ledger");
                    FilesUtil.write(work.resolve(action + "-done.json"), FilesUtil.json(Map.of("conflicts", conflicts, "replay_stable", true, "attempts", attempts)));
                    return;
                } catch (HeadConflictException conflict) {
                    conflicts++;
                    head = client.query.head(version);
                    if (head != 1) throw new IllegalStateException("Unexpected competing head");
                }
            }
            throw new IllegalStateException("Worker did not finish within retry bound");
        }
    }
}

File: src/shift-handover/src/competing_workers.rs
// SPDX-License-Identifier: Apache-2.0
use crate::{bootstrap, data};
use anyhow::{ensure, Result};
use munarium_client::{dto, MunariumError};
use serde_json::json;
use std::{path::Path, time::Duration};

async fn barrier(path: impl AsRef<Path>) -> Result<()> {
    let deadline = tokio::time::Instant::now() + Duration::from_secs(60);
    while !path.as_ref().exists() {
        ensure!(
            tokio::time::Instant::now() < deadline,
            "Worker barrier timed out"
        );
        tokio::time::sleep(Duration::from_millis(50)).await;
    }
    Ok(())
}

pub async fn run(action: &str, work: &str) -> Result<()> {
    let c = bootstrap::writer_at(&bootstrap::endpoint())?;
    let path = |name: &str| Path::new(work).join(name);
    if action == "init" {
        ensure!(!path("version.json").exists(), "Fresh race state required");
        let result = c
            .commands
            .create_version(
                dto::CreateVersionRequest {
                    parent_version_id: None,
                    metadata: Some(json!({"application":"shift-worker-race"})),
                },
                Some(uuid::Uuid::new_v4().to_string()),
            )
            .await?;
        data::save(path("version.json"), &json!({"version":result.version_id}))?;
        return Ok(());
    }
    let state = data::read(path("version.json"))?;
    let version = state["version"]
        .as_str()
        .ok_or_else(|| anyhow::anyhow!("Missing version"))?;
    if action == "verify" {
        let a = data::read(path("a-done.json"))?;
        let b = data::read(path("b-done.json"))?;
        let conflicts =
            a["conflicts"].as_u64().unwrap_or(99) + b["conflicts"].as_u64().unwrap_or(99);
        ensure!(
            conflicts == 1
                && c.query.head(version).await? == 2
                && a["replay_stable"] == true
                && b["replay_stable"] == true,
            "Expected one conflict, two writes and stable completed replays"
        );
        data::save(
            path("quality.json"),
            &json!({"passed":true,"tests":1,"workers":2,"conflicts":conflicts,"version":version}),
        )?;
        data::write(path("tests.xml"), "<testsuite name=\"competing-worker-containers\" tests=\"1\" failures=\"0\" skipped=\"0\"><testcase name=\"fresh-head-and-key-after-conflict\"/></testsuite>")?;
        return Ok(());
    }
    ensure!(action == "a" || action == "b", "Unknown worker");
    let peer = if action == "a" { "b" } else { "a" };
    let mut head = c.query.head(version).await?;
    ensure!(head == 0, "Fresh ledger required");
    data::save(path(&format!("{action}-ready.json")), &json!({"head":head}))?;
    barrier(path(&format!("{peer}-ready.json"))).await?;
    let mut attempts = Vec::new();
    let mut conflicts = 0;
    for _ in 0..2 {
        let key = uuid::Uuid::new_v4().to_string();
        let body = json!({"expected_head":head,"claim_type":"fact","subject":format!("worker_{action}"),"key":"handover_review","value":"reviewed"});
        attempts.push(json!({"idempotency_key":key,"body":body}));
        data::save(path(&format!("{action}-journal.json")), &attempts)?;
        match c
            .commands
            .propose_claim(
                version,
                serde_json::from_value(body.clone())?,
                Some(key.clone()),
            )
            .await
        {
            Ok(result) => {
                let accepted = serde_json::to_value(result)?;
                ensure!(
                    accepted["claim"]["id"].is_string(),
                    "Missing accepted claim identity"
                );
                data::save(path(&format!("{action}-accepted.json")), &accepted)?;
                barrier(path(&format!("{peer}-accepted.json"))).await?;
                let completed_head = c.query.head(version).await?;
                let replay = serde_json::to_value(
                    c.commands
                        .propose_claim(version, serde_json::from_value(body)?, Some(key))
                        .await?,
                )?;
                ensure!(
                    replay["claim"]["id"] == accepted["claim"]["id"]
                        && c.query.head(version).await? == completed_head,
                    "Completed replay changed ledger"
                );
                data::save(
                    path(&format!("{action}-done.json")),
                    &json!({"conflicts":conflicts,"replay_stable":true,"attempts":attempts}),
                )?;
                return Ok(());
            }
            Err(MunariumError::HeadConflict { .. }) => {
                conflicts += 1;
                head = c.query.head(version).await?;
                ensure!(head == 1, "Unexpected competing head");
            }
            Err(error) => return Err(error.into()),
        }
    }
    anyhow::bail!("Worker did not finish within retry bound")
}
